jupyter-org-client.el

```diff
@@ -416,6 +416,357 @@ the `syntax-table' will be set to that of the REPL buffers."
+                                      (org-element-property :end src-block))
+                    ""))
+            (list next-src-block current-src-block))
+      ;; Now create the merged block, point is where the current block was
+      (insert (format "#+BEGIN_SRC %s %s %s\n%s#+END_SRC\n\n"
+                      lang switches parameters merged-code))
+      (forward-line -3)
+      (end-of-line))))
+
+;;;###autoload
+(defun jupyter-org-move-src-block (&optional below)
+  "Move source block before of after another.
+
+If BELOW is non-nil, move the block down, otherwise move it up."
+  (interactive)
+  ;;skip all this if there are no previous or next source block
+  (when (condition-case nil
+            (save-excursion
+              (if below
+                  (org-babel-next-src-block)
+                (org-babel-previous-src-block)))
+          (error nil))
+    (let* ((src (org-element-context))
+           (results-start (org-babel-where-is-src-block-result))
+           (results-end
+            (when results-start
+              (save-excursion
+                (goto-char results-start)
+                (goto-char (org-babel-result-end))
+                ;; if line is empty, take that empty line by moving down
+                (when (looking-at-p "[[:space:]]*$")
+                  (forward-line 1))
+                (point)))))
+      ;; kill from the start of the source block to the end of the results
+      ;; ... if there are no results, to the end of the source block
+      (kill-region
+       (org-element-property :begin src)
+       (or results-end (org-element-property :end src))))
+    (if below
+        ;; if below, move past the next source block or its result
+        (let ((next-src-block-head (org-babel-where-is-src-block-head)))
+          (if next-src-block-head
+              (goto-char next-src-block-head)
+            (org-babel-next-src-block))
+          (let ((next-src-block (org-element-context))
+                (next-results-start (org-babel-where-is-src-block-result)))
+            (if next-results-start
+                (progn
+                  (goto-char next-results-start)
+                  (goto-char (org-babel-result-end))
+                  (when (looking-at-p "[[:space:]]*$")
+                    (forward-line 1)))
+              (goto-char (org-element-property :end next-src-block)))))
+      ;; else, move to the begining of the previous block
+      (org-babel-previous-src-block))
+    (save-excursion (org-yank)))) ; keep cursor where the yank takes place
+
+;;;###autoload
+(defun jupyter-org-clear-all-results ()
+  "Clear all results in the buffer."
+  (interactive)
+  (save-excursion
+    (goto-char (point-min))
+    (while (org-babel-next-src-block)
+      (org-babel-remove-result))))
+
+;;;###autoload
+(defun jupyter-org-restart-kernel-execute-block ()
+  "Restart kernel and execute block."
+  (interactive)
+  (let* ((params (car jupyter-org--src-block-cache))
```

Comment:
This looks like a job for `jupyter-org-with-src-block-client`.

Reply:
With the following code all seems to work ok
```lisp
(jupyter-org-with-src-block-client
   (jupyter-repl-restart-kernel))
```
However, no matter how I restart the kernel, I get some messages in the repl's stderr. The buffer name of the stderr is "*jupyter-*jupyter-repl[python 3.7.2]-testpy2-stderr**" and contains:
```
ERROR:root:Invalid alias: The name clear can't be aliased because it is another
magic command.
ERROR:root:Invalid alias: The name more can't be aliased because it is another
magic command.
ERROR:root:Invalid alias: The name less can't be aliased because it is another
magic command.
ERROR:root:Invalid alias: The name man can't be aliased because it is another
magic command.
```
And the following error message:
```log
Restarting kernel...
Error running timer ‘jupyter-handle-message’: (error "Found end of previous cell")
python3 kernel shutting down...done
Starting python3 kernel process...done
```

Reply:
I don't know about all of the Python errors, seems like more of a Python issue. I am on Python 3.7.1 and I haven't been getting those errors when I testing the kernel restart stuff. Can you try 1f5c7743d534929efaad7592762f241700bc1f6c to see if that restarts the kernel more cleanly.

Reply:
With the latest master, the kernel restarts more cleanly (less error messages) but there are still some error messages:
```
Error running timer ‘jupyter-handle-message’: (error "Cell for request not found") [4 times]
```
The trace I get for the error is this one:
```lisp
Debugger entered--Lisp error: (error "Cell for request not found")
  signal(error ("Cell for request not found"))
  error("Cell for request not found")
  jupyter-repl-goto-cell(#s(jupyter-request :id "cfd3b78d-1ec6-409b-ba97-41c2cb82c655" :time (23677 1564 171792 52000) :idle-received-p t :last-message (:header (message-part "{\"msg_id\":\"e8c6a08d-83b1c9f7d74a3dfb1b80248e\",\"msg_type\":\"stream\",\"username\":\"boy\",\"session\":\"464ef51b-d62e8feaf3cf3b8548035101\",\"date\":\"2019-03-04T11:03:56.326894Z\",\"version\":\"5.3\"}" (:msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :username "boy" :session "464ef51b-d62e8feaf3cf3b8548035101" :date "2019-03-04T11:03:56.326894Z" :version "5.3")) :msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :parent_header (message-part "{\"msg_id\":\"cfd3b78d-1ec6-409b-ba97-41c2cb82c655\",\"msg_type\":\"shutdown_request\",\"version\":\"5.3\",\"username\":\"boy\",\"session\":\"9b3dd7de-5a5f-4878-8643-a179c8a4bee2\",\"date\":\"2019-03-04T12:03:56.171874+01:00\"}" (:msg_id "cfd3b78d-1ec6-409b-ba97-41c2cb82c655" :msg_type :shutdown-request :version "5.3" :username "boy" :session "9b3dd7de-5a5f-4878-8643-a179c8a4bee2" :date "2019-03-04T12:03:56.171874+01:00")) :metadata (message-part "{}" nil) :content (message-part "{\"name\":\"stderr\",\"text\":\"ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\\n\"}" (:name "stderr" :text "ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\n")) :buffers nil) :inhibited-handlers nil :callbacks ((:shutdown-reply . #f(compiled-function (msg) #<bytecode 0x1240cc1>)))))
  #f(compiled-function (client req name text) #<bytecode 0x23c09d5>)(#<jupyter-org-client jupyter-org-client> #s(jupyter-request :id "cfd3b78d-1ec6-409b-ba97-41c2cb82c655" :time (23677 1564 171792 52000) :idle-received-p t :last-message (:header (message-part "{\"msg_id\":\"e8c6a08d-83b1c9f7d74a3dfb1b80248e\",\"msg_type\":\"stream\",\"username\":\"boy\",\"session\":\"464ef51b-d62e8feaf3cf3b8548035101\",\"date\":\"2019-03-04T11:03:56.326894Z\",\"version\":\"5.3\"}" (:msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :username "boy" :session "464ef51b-d62e8feaf3cf3b8548035101" :date "2019-03-04T11:03:56.326894Z" :version "5.3")) :msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :parent_header (message-part "{\"msg_id\":\"cfd3b78d-1ec6-409b-ba97-41c2cb82c655\",\"msg_type\":\"shutdown_request\",\"version\":\"5.3\",\"username\":\"boy\",\"session\":\"9b3dd7de-5a5f-4878-8643-a179c8a4bee2\",\"date\":\"2019-03-04T12:03:56.171874+01:00\"}" (:msg_id "cfd3b78d-1ec6-409b-ba97-41c2cb82c655" :msg_type :shutdown-request :version "5.3" :username "boy" :session "9b3dd7de-5a5f-4878-8643-a179c8a4bee2" :date "2019-03-04T12:03:56.171874+01:00")) :metadata (message-part "{}" nil) :content (message-part "{\"name\":\"stderr\",\"text\":\"ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\\n\"}" (:name "stderr" :text "ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\n")) :buffers nil) :inhibited-handlers nil :callbacks ((:shutdown-reply . #f(compiled-function (msg) #<bytecode 0x1240cc1>)))) "stderr" "ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\n")
  apply(#f(compiled-function (client req name text) #<bytecode 0x23c09d5>) #<jupyter-org-client jupyter-org-client> (#s(jupyter-request :id "cfd3b78d-1ec6-409b-ba97-41c2cb82c655" :time (23677 1564 171792 52000) :idle-received-p t :last-message (:header (message-part "{\"msg_id\":\"e8c6a08d-83b1c9f7d74a3dfb1b80248e\",\"msg_type\":\"stream\",\"username\":\"boy\",\"session\":\"464ef51b-d62e8feaf3cf3b8548035101\",\"date\":\"2019-03-04T11:03:56.326894Z\",\"version\":\"5.3\"}" (:msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :username "boy" :session "464ef51b-d62e8feaf3cf3b8548035101" :date "2019-03-04T11:03:56.326894Z" :version "5.3")) :msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :parent_header (message-part "{\"msg_id\":\"cfd3b78d-1ec6-409b-ba97-41c2cb82c655\",\"msg_type\":\"shutdown_request\",\"version\":\"5.3\",\"username\":\"boy\",\"session\":\"9b3dd7de-5a5f-4878-8643-a179c8a4bee2\",\"date\":\"2019-03-04T12:03:56.171874+01:00\"}" (:msg_id "cfd3b78d-1ec6-409b-ba97-41c2cb82c655" :msg_type :shutdown-request :version "5.3" :username "boy" :session "9b3dd7de-5a5f-4878-8643-a179c8a4bee2" :date "2019-03-04T12:03:56.171874+01:00")) :metadata (message-part "{}" nil) :content (message-part "{\"name\":\"stderr\",\"text\":\"ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\\n\"}" (:name "stderr" :text "ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\n")) :buffers nil) :inhibited-handlers nil :callbacks ((:shutdown-reply . #f(compiled-function (msg) #<bytecode 0x1240cc1>)))) "stderr" "ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\n"))
  #f(compiled-function (arg &rest args) #<bytecode 0x12d69c5>)(#<jupyter-org-client jupyter-org-client> #s(jupyter-request :id "cfd3b78d-1ec6-409b-ba97-41c2cb82c655" :time (23677 1564 171792 52000) :idle-received-p t :last-message (:header (message-part "{\"msg_id\":\"e8c6a08d-83b1c9f7d74a3dfb1b80248e\",\"msg_type\":\"stream\",\"username\":\"boy\",\"session\":\"464ef51b-d62e8feaf3cf3b8548035101\",\"date\":\"2019-03-04T11:03:56.326894Z\",\"version\":\"5.3\"}" (:msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :username "boy" :session "464ef51b-d62e8feaf3cf3b8548035101" :date "2019-03-04T11:03:56.326894Z" :version "5.3")) :msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :parent_header (message-part "{\"msg_id\":\"cfd3b78d-1ec6-409b-ba97-41c2cb82c655\",\"msg_type\":\"shutdown_request\",\"version\":\"5.3\",\"username\":\"boy\",\"session\":\"9b3dd7de-5a5f-4878-8643-a179c8a4bee2\",\"date\":\"2019-03-04T12:03:56.171874+01:00\"}" (:msg_id "cfd3b78d-1ec6-409b-ba97-41c2cb82c655" :msg_type :shutdown-request :version "5.3" :username "boy" :session "9b3dd7de-5a5f-4878-8643-a179c8a4bee2" :date "2019-03-04T12:03:56.171874+01:00")) :metadata (message-part "{}" nil) :content (message-part "{\"name\":\"stderr\",\"text\":\"ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\\n\"}" (:name "stderr" :text "ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\n")) :buffers nil) :inhibited-handlers nil :callbacks ((:shutdown-reply . #f(compiled-function (msg) #<bytecode 0x1240cc1>)))) "stderr" "ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\n")
  apply(#f(compiled-function (arg &rest args) #<bytecode 0x12d69c5>) #<jupyter-org-client jupyter-org-client> #s(jupyter-request :id "cfd3b78d-1ec6-409b-ba97-41c2cb82c655" :time (23677 1564 171792 52000) :idle-received-p t :last-message (:header (message-part "{\"msg_id\":\"e8c6a08d-83b1c9f7d74a3dfb1b80248e\",\"msg_type\":\"stream\",\"username\":\"boy\",\"session\":\"464ef51b-d62e8feaf3cf3b8548035101\",\"date\":\"2019-03-04T11:03:56.326894Z\",\"version\":\"5.3\"}" (:msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :username "boy" :session "464ef51b-d62e8feaf3cf3b8548035101" :date "2019-03-04T11:03:56.326894Z" :version "5.3")) :msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :parent_header (message-part "{\"msg_id\":\"cfd3b78d-1ec6-409b-ba97-41c2cb82c655\",\"msg_type\":\"shutdown_request\",\"version\":\"5.3\",\"username\":\"boy\",\"session\":\"9b3dd7de-5a5f-4878-8643-a179c8a4bee2\",\"date\":\"2019-03-04T12:03:56.171874+01:00\"}" (:msg_id "cfd3b78d-1ec6-409b-ba97-41c2cb82c655" :msg_type :shutdown-request :version "5.3" :username "boy" :session "9b3dd7de-5a5f-4878-8643-a179c8a4bee2" :date "2019-03-04T12:03:56.171874+01:00")) :metadata (message-part "{}" nil) :content (message-part "{\"name\":\"stderr\",\"text\":\"ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\\n\"}" (:name "stderr" :text "ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\n")) :buffers nil) :inhibited-handlers nil :callbacks ((:shutdown-reply . #f(compiled-function (msg) #<bytecode 0x1240cc1>)))) ("stderr" "ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\n"))
  jupyter-handle-stream(#<jupyter-org-client jupyter-org-client> #s(jupyter-request :id "cfd3b78d-1ec6-409b-ba97-41c2cb82c655" :time (23677 1564 171792 52000) :idle-received-p t :last-message (:header (message-part "{\"msg_id\":\"e8c6a08d-83b1c9f7d74a3dfb1b80248e\",\"msg_type\":\"stream\",\"username\":\"boy\",\"session\":\"464ef51b-d62e8feaf3cf3b8548035101\",\"date\":\"2019-03-04T11:03:56.326894Z\",\"version\":\"5.3\"}" (:msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :username "boy" :session "464ef51b-d62e8feaf3cf3b8548035101" :date "2019-03-04T11:03:56.326894Z" :version "5.3")) :msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :parent_header (message-part "{\"msg_id\":\"cfd3b78d-1ec6-409b-ba97-41c2cb82c655\",\"msg_type\":\"shutdown_request\",\"version\":\"5.3\",\"username\":\"boy\",\"session\":\"9b3dd7de-5a5f-4878-8643-a179c8a4bee2\",\"date\":\"2019-03-04T12:03:56.171874+01:00\"}" (:msg_id "cfd3b78d-1ec6-409b-ba97-41c2cb82c655" :msg_type :shutdown-request :version "5.3" :username "boy" :session "9b3dd7de-5a5f-4878-8643-a179c8a4bee2" :date "2019-03-04T12:03:56.171874+01:00")) :metadata (message-part "{}" nil) :content (message-part "{\"name\":\"stderr\",\"text\":\"ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\\n\"}" (:name "stderr" :text "ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\n")) :buffers nil) :inhibited-handlers nil :callbacks ((:shutdown-reply . #f(compiled-function (msg) #<bytecode 0x1240cc1>)))) "stderr" "ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\n")
  #f(compiled-function (channel client req msg) #<bytecode 0x232c895>)(:iopub #<jupyter-org-client jupyter-org-client> #s(jupyter-request :id "cfd3b78d-1ec6-409b-ba97-41c2cb82c655" :time (23677 1564 171792 52000) :idle-received-p t :last-message (:header (message-part "{\"msg_id\":\"e8c6a08d-83b1c9f7d74a3dfb1b80248e\",\"msg_type\":\"stream\",\"username\":\"boy\",\"session\":\"464ef51b-d62e8feaf3cf3b8548035101\",\"date\":\"2019-03-04T11:03:56.326894Z\",\"version\":\"5.3\"}" (:msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :username "boy" :session "464ef51b-d62e8feaf3cf3b8548035101" :date "2019-03-04T11:03:56.326894Z" :version "5.3")) :msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :parent_header (message-part "{\"msg_id\":\"cfd3b78d-1ec6-409b-ba97-41c2cb82c655\",\"msg_type\":\"shutdown_request\",\"version\":\"5.3\",\"username\":\"boy\",\"session\":\"9b3dd7de-5a5f-4878-8643-a179c8a4bee2\",\"date\":\"2019-03-04T12:03:56.171874+01:00\"}" (:msg_id "cfd3b78d-1ec6-409b-ba97-41c2cb82c655" :msg_type :shutdown-request :version "5.3" :username "boy" :session "9b3dd7de-5a5f-4878-8643-a179c8a4bee2" :date "2019-03-04T12:03:56.171874+01:00")) :metadata (message-part "{}" nil) :content (message-part "{\"name\":\"stderr\",\"text\":\"ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\\n\"}" (:name "stderr" :text "ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\n")) :buffers nil) :inhibited-handlers nil :callbacks ((:shutdown-reply . #f(compiled-function (msg) #<bytecode 0x1240cc1>)))) (:header (message-part "{\"msg_id\":\"e8c6a08d-83b1c9f7d74a3dfb1b80248e\",\"msg_type\":\"stream\",\"username\":\"boy\",\"session\":\"464ef51b-d62e8feaf3cf3b8548035101\",\"date\":\"2019-03-04T11:03:56.326894Z\",\"version\":\"5.3\"}" (:msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :username "boy" :session "464ef51b-d62e8feaf3cf3b8548035101" :date "2019-03-04T11:03:56.326894Z" :version "5.3")) :msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :parent_header (message-part "{\"msg_id\":\"cfd3b78d-1ec6-409b-ba97-41c2cb82c655\",\"msg_type\":\"shutdown_request\",\"version\":\"5.3\",\"username\":\"boy\",\"session\":\"9b3dd7de-5a5f-4878-8643-a179c8a4bee2\",\"date\":\"2019-03-04T12:03:56.171874+01:00\"}" (:msg_id "cfd3b78d-1ec6-409b-ba97-41c2cb82c655" :msg_type :shutdown-request :version "5.3" :username "boy" :session "9b3dd7de-5a5f-4878-8643-a179c8a4bee2" :date "2019-03-04T12:03:56.171874+01:00")) :metadata (message-part "{}" nil) :content (message-part "{\"name\":\"stderr\",\"text\":\"ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\\n\"}" (:name "stderr" :text "ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\n")) :buffers nil))
  apply(#f(compiled-function (channel client req msg) #<bytecode 0x232c895>) :iopub (#<jupyter-org-client jupyter-org-client> #s(jupyter-request :id "cfd3b78d-1ec6-409b-ba97-41c2cb82c655" :time (23677 1564 171792 52000) :idle-received-p t :last-message (:header (message-part "{\"msg_id\":\"e8c6a08d-83b1c9f7d74a3dfb1b80248e\",\"msg_type\":\"stream\",\"username\":\"boy\",\"session\":\"464ef51b-d62e8feaf3cf3b8548035101\",\"date\":\"2019-03-04T11:03:56.326894Z\",\"version\":\"5.3\"}" (:msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :username "boy" :session "464ef51b-d62e8feaf3cf3b8548035101" :date "2019-03-04T11:03:56.326894Z" :version "5.3")) :msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :parent_header (message-part "{\"msg_id\":\"cfd3b78d-1ec6-409b-ba97-41c2cb82c655\",\"msg_type\":\"shutdown_request\",\"version\":\"5.3\",\"username\":\"boy\",\"session\":\"9b3dd7de-5a5f-4878-8643-a179c8a4bee2\",\"date\":\"2019-03-04T12:03:56.171874+01:00\"}" (:msg_id "cfd3b78d-1ec6-409b-ba97-41c2cb82c655" :msg_type :shutdown-request :version "5.3" :username "boy" :session "9b3dd7de-5a5f-4878-8643-a179c8a4bee2" :date "2019-03-04T12:03:56.171874+01:00")) :metadata (message-part "{}" nil) :content (message-part "{\"name\":\"stderr\",\"text\":\"ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\\n\"}" (:name "stderr" :text "ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\n")) :buffers nil) :inhibited-handlers nil :callbacks ((:shutdown-reply . #f(compiled-function (msg) #<bytecode 0x1240cc1>)))) (:header (message-part "{\"msg_id\":\"e8c6a08d-83b1c9f7d74a3dfb1b80248e\",\"msg_type\":\"stream\",\"username\":\"boy\",\"session\":\"464ef51b-d62e8feaf3cf3b8548035101\",\"date\":\"2019-03-04T11:03:56.326894Z\",\"version\":\"5.3\"}" (:msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :username "boy" :session "464ef51b-d62e8feaf3cf3b8548035101" :date "2019-03-04T11:03:56.326894Z" :version "5.3")) :msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :parent_header (message-part "{\"msg_id\":\"cfd3b78d-1ec6-409b-ba97-41c2cb82c655\",\"msg_type\":\"shutdown_request\",\"version\":\"5.3\",\"username\":\"boy\",\"session\":\"9b3dd7de-5a5f-4878-8643-a179c8a4bee2\",\"date\":\"2019-03-04T12:03:56.171874+01:00\"}" (:msg_id "cfd3b78d-1ec6-409b-ba97-41c2cb82c655" :msg_type :shutdown-request :version "5.3" :username "boy" :session "9b3dd7de-5a5f-4878-8643-a179c8a4bee2" :date "2019-03-04T12:03:56.171874+01:00")) :metadata (message-part "{}" nil) :content (message-part "{\"name\":\"stderr\",\"text\":\"ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\\n\"}" (:name "stderr" :text "ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\n")) :buffers nil)))
  jupyter-handle-message(:iopub #<jupyter-org-client jupyter-org-client> #s(jupyter-request :id "cfd3b78d-1ec6-409b-ba97-41c2cb82c655" :time (23677 1564 171792 52000) :idle-received-p t :last-message (:header (message-part "{\"msg_id\":\"e8c6a08d-83b1c9f7d74a3dfb1b80248e\",\"msg_type\":\"stream\",\"username\":\"boy\",\"session\":\"464ef51b-d62e8feaf3cf3b8548035101\",\"date\":\"2019-03-04T11:03:56.326894Z\",\"version\":\"5.3\"}" (:msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :username "boy" :session "464ef51b-d62e8feaf3cf3b8548035101" :date "2019-03-04T11:03:56.326894Z" :version "5.3")) :msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :parent_header (message-part "{\"msg_id\":\"cfd3b78d-1ec6-409b-ba97-41c2cb82c655\",\"msg_type\":\"shutdown_request\",\"version\":\"5.3\",\"username\":\"boy\",\"session\":\"9b3dd7de-5a5f-4878-8643-a179c8a4bee2\",\"date\":\"2019-03-04T12:03:56.171874+01:00\"}" (:msg_id "cfd3b78d-1ec6-409b-ba97-41c2cb82c655" :msg_type :shutdown-request :version "5.3" :username "boy" :session "9b3dd7de-5a5f-4878-8643-a179c8a4bee2" :date "2019-03-04T12:03:56.171874+01:00")) :metadata (message-part "{}" nil) :content (message-part "{\"name\":\"stderr\",\"text\":\"ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\\n\"}" (:name "stderr" :text "ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\n")) :buffers nil) :inhibited-handlers nil :callbacks ((:shutdown-reply . #f(compiled-function (msg) #<bytecode 0x1240cc1>)))) (:header (message-part "{\"msg_id\":\"e8c6a08d-83b1c9f7d74a3dfb1b80248e\",\"msg_type\":\"stream\",\"username\":\"boy\",\"session\":\"464ef51b-d62e8feaf3cf3b8548035101\",\"date\":\"2019-03-04T11:03:56.326894Z\",\"version\":\"5.3\"}" (:msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :username "boy" :session "464ef51b-d62e8feaf3cf3b8548035101" :date "2019-03-04T11:03:56.326894Z" :version "5.3")) :msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :parent_header (message-part "{\"msg_id\":\"cfd3b78d-1ec6-409b-ba97-41c2cb82c655\",\"msg_type\":\"shutdown_request\",\"version\":\"5.3\",\"username\":\"boy\",\"session\":\"9b3dd7de-5a5f-4878-8643-a179c8a4bee2\",\"date\":\"2019-03-04T12:03:56.171874+01:00\"}" (:msg_id "cfd3b78d-1ec6-409b-ba97-41c2cb82c655" :msg_type :shutdown-request :version "5.3" :username "boy" :session "9b3dd7de-5a5f-4878-8643-a179c8a4bee2" :date "2019-03-04T12:03:56.171874+01:00")) :metadata (message-part "{}" nil) :content (message-part "{\"name\":\"stderr\",\"text\":\"ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\\n\"}" (:name "stderr" :text "ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\n")) :buffers nil))
  jupyter--run-handler-maybe(#<jupyter-org-client jupyter-org-client> :iopub #s(jupyter-request :id "cfd3b78d-1ec6-409b-ba97-41c2cb82c655" :time (23677 1564 171792 52000) :idle-received-p t :last-message (:header (message-part "{\"msg_id\":\"e8c6a08d-83b1c9f7d74a3dfb1b80248e\",\"msg_type\":\"stream\",\"username\":\"boy\",\"session\":\"464ef51b-d62e8feaf3cf3b8548035101\",\"date\":\"2019-03-04T11:03:56.326894Z\",\"version\":\"5.3\"}" (:msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :username "boy" :session "464ef51b-d62e8feaf3cf3b8548035101" :date "2019-03-04T11:03:56.326894Z" :version "5.3")) :msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :parent_header (message-part "{\"msg_id\":\"cfd3b78d-1ec6-409b-ba97-41c2cb82c655\",\"msg_type\":\"shutdown_request\",\"version\":\"5.3\",\"username\":\"boy\",\"session\":\"9b3dd7de-5a5f-4878-8643-a179c8a4bee2\",\"date\":\"2019-03-04T12:03:56.171874+01:00\"}" (:msg_id "cfd3b78d-1ec6-409b-ba97-41c2cb82c655" :msg_type :shutdown-request :version "5.3" :username "boy" :session "9b3dd7de-5a5f-4878-8643-a179c8a4bee2" :date "2019-03-04T12:03:56.171874+01:00")) :metadata (message-part "{}" nil) :content (message-part "{\"name\":\"stderr\",\"text\":\"ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\\n\"}" (:name "stderr" :text "ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\n")) :buffers nil) :inhibited-handlers nil :callbacks ((:shutdown-reply . #f(compiled-function (msg) #<bytecode 0x1240cc1>)))) (:header (message-part "{\"msg_id\":\"e8c6a08d-83b1c9f7d74a3dfb1b80248e\",\"msg_type\":\"stream\",\"username\":\"boy\",\"session\":\"464ef51b-d62e8feaf3cf3b8548035101\",\"date\":\"2019-03-04T11:03:56.326894Z\",\"version\":\"5.3\"}" (:msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :username "boy" :session "464ef51b-d62e8feaf3cf3b8548035101" :date "2019-03-04T11:03:56.326894Z" :version "5.3")) :msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :parent_header (message-part "{\"msg_id\":\"cfd3b78d-1ec6-409b-ba97-41c2cb82c655\",\"msg_type\":\"shutdown_request\",\"version\":\"5.3\",\"username\":\"boy\",\"session\":\"9b3dd7de-5a5f-4878-8643-a179c8a4bee2\",\"date\":\"2019-03-04T12:03:56.171874+01:00\"}" (:msg_id "cfd3b78d-1ec6-409b-ba97-41c2cb82c655" :msg_type :shutdown-request :version "5.3" :username "boy" :session "9b3dd7de-5a5f-4878-8643-a179c8a4bee2" :date "2019-03-04T12:03:56.171874+01:00")) :metadata (message-part "{}" nil) :content (message-part "{\"name\":\"stderr\",\"text\":\"ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\\n\"}" (:name "stderr" :text "ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\n")) :buffers nil))
  #f(compiled-function () #<bytecode 0x1bfbe61>)()
  #f(compiled-function (client channel msg) "Process a message on CLIENT's CHANNEL.\nWhen a message is received from the kernel, the\n`jupyter-handle-message' method is called on the client. The\nclient method runs any callbacks for the message and possibly\nruns the client handler for the channel the message was received\non. The channel's `jupyter-handle-message' method will then pass\nthe message to the appropriate message handler based on the\nmessage type.\n\nSo when a message is received from the kernel the following steps\nare taken:\n\n- `jupyter-handle-message' (client)\n   - Run callbacks for message\n   - Possibly run channel handlers\n     - `jupyter-handle-message' (channel)\n       - Based on message type, dispatch to\n         `jupyter-handle-execute-result',\n         `jupyter-handle-kernel-info-reply', ...\n   - Remove request from client request table when idle message is received" #<bytecode 0x2318a99>)(#<jupyter-org-client jupyter-org-client> :iopub (:header (message-part "{\"msg_id\":\"e8c6a08d-83b1c9f7d74a3dfb1b80248e\",\"msg_type\":\"stream\",\"username\":\"boy\",\"session\":\"464ef51b-d62e8feaf3cf3b8548035101\",\"date\":\"2019-03-04T11:03:56.326894Z\",\"version\":\"5.3\"}" (:msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :username "boy" :session "464ef51b-d62e8feaf3cf3b8548035101" :date "2019-03-04T11:03:56.326894Z" :version "5.3")) :msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :parent_header (message-part "{\"msg_id\":\"cfd3b78d-1ec6-409b-ba97-41c2cb82c655\",\"msg_type\":\"shutdown_request\",\"version\":\"5.3\",\"username\":\"boy\",\"session\":\"9b3dd7de-5a5f-4878-8643-a179c8a4bee2\",\"date\":\"2019-03-04T12:03:56.171874+01:00\"}" (:msg_id "cfd3b78d-1ec6-409b-ba97-41c2cb82c655" :msg_type :shutdown-request :version "5.3" :username "boy" :session "9b3dd7de-5a5f-4878-8643-a179c8a4bee2" :date "2019-03-04T12:03:56.171874+01:00")) :metadata (message-part "{}" nil) :content (message-part "{\"name\":\"stderr\",\"text\":\"ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\\n\"}" (:name "stderr" :text "ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\n")) :buffers nil))
  apply(#f(compiled-function (client channel msg) "Process a message on CLIENT's CHANNEL.\nWhen a message is received from the kernel, the\n`jupyter-handle-message' method is called on the client. The\nclient method runs any callbacks for the message and possibly\nruns the client handler for the channel the message was received\non. The channel's `jupyter-handle-message' method will then pass\nthe message to the appropriate message handler based on the\nmessage type.\n\nSo when a message is received from the kernel the following steps\nare taken:\n\n- `jupyter-handle-message' (client)\n   - Run callbacks for message\n   - Possibly run channel handlers\n     - `jupyter-handle-message' (channel)\n       - Based on message type, dispatch to\n         `jupyter-handle-execute-result',\n         `jupyter-handle-kernel-info-reply', ...\n   - Remove request from client request table when idle message is received" #<bytecode 0x2318a99>) #<jupyter-org-client jupyter-org-client> (:iopub (:header (message-part "{\"msg_id\":\"e8c6a08d-83b1c9f7d74a3dfb1b80248e\",\"msg_type\":\"stream\",\"username\":\"boy\",\"session\":\"464ef51b-d62e8feaf3cf3b8548035101\",\"date\":\"2019-03-04T11:03:56.326894Z\",\"version\":\"5.3\"}" (:msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :username "boy" :session "464ef51b-d62e8feaf3cf3b8548035101" :date "2019-03-04T11:03:56.326894Z" :version "5.3")) :msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :parent_header (message-part "{\"msg_id\":\"cfd3b78d-1ec6-409b-ba97-41c2cb82c655\",\"msg_type\":\"shutdown_request\",\"version\":\"5.3\",\"username\":\"boy\",\"session\":\"9b3dd7de-5a5f-4878-8643-a179c8a4bee2\",\"date\":\"2019-03-04T12:03:56.171874+01:00\"}" (:msg_id "cfd3b78d-1ec6-409b-ba97-41c2cb82c655" :msg_type :shutdown-request :version "5.3" :username "boy" :session "9b3dd7de-5a5f-4878-8643-a179c8a4bee2" :date "2019-03-04T12:03:56.171874+01:00")) :metadata (message-part "{}" nil) :content (message-part "{\"name\":\"stderr\",\"text\":\"ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\\n\"}" (:name "stderr" :text "ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\n")) :buffers nil)))
  jupyter-handle-message(#<jupyter-org-client jupyter-org-client> :iopub (:header (message-part "{\"msg_id\":\"e8c6a08d-83b1c9f7d74a3dfb1b80248e\",\"msg_type\":\"stream\",\"username\":\"boy\",\"session\":\"464ef51b-d62e8feaf3cf3b8548035101\",\"date\":\"2019-03-04T11:03:56.326894Z\",\"version\":\"5.3\"}" (:msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :username "boy" :session "464ef51b-d62e8feaf3cf3b8548035101" :date "2019-03-04T11:03:56.326894Z" :version "5.3")) :msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :parent_header (message-part "{\"msg_id\":\"cfd3b78d-1ec6-409b-ba97-41c2cb82c655\",\"msg_type\":\"shutdown_request\",\"version\":\"5.3\",\"username\":\"boy\",\"session\":\"9b3dd7de-5a5f-4878-8643-a179c8a4bee2\",\"date\":\"2019-03-04T12:03:56.171874+01:00\"}" (:msg_id "cfd3b78d-1ec6-409b-ba97-41c2cb82c655" :msg_type :shutdown-request :version "5.3" :username "boy" :session "9b3dd7de-5a5f-4878-8643-a179c8a4bee2" :date "2019-03-04T12:03:56.171874+01:00")) :metadata (message-part "{}" nil) :content (message-part "{\"name\":\"stderr\",\"text\":\"ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\\n\"}" (:name "stderr" :text "ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\n")) :buffers nil))
  apply(jupyter-handle-message (#<jupyter-org-client jupyter-org-client> :iopub (:header (message-part "{\"msg_id\":\"e8c6a08d-83b1c9f7d74a3dfb1b80248e\",\"msg_type\":\"stream\",\"username\":\"boy\",\"session\":\"464ef51b-d62e8feaf3cf3b8548035101\",\"date\":\"2019-03-04T11:03:56.326894Z\",\"version\":\"5.3\"}" (:msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :username "boy" :session "464ef51b-d62e8feaf3cf3b8548035101" :date "2019-03-04T11:03:56.326894Z" :version "5.3")) :msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :parent_header (message-part "{\"msg_id\":\"cfd3b78d-1ec6-409b-ba97-41c2cb82c655\",\"msg_type\":\"shutdown_request\",\"version\":\"5.3\",\"username\":\"boy\",\"session\":\"9b3dd7de-5a5f-4878-8643-a179c8a4bee2\",\"date\":\"2019-03-04T12:03:56.171874+01:00\"}" (:msg_id "cfd3b78d-1ec6-409b-ba97-41c2cb82c655" :msg_type :shutdown-request :version "5.3" :username "boy" :session "9b3dd7de-5a5f-4878-8643-a179c8a4bee2" :date "2019-03-04T12:03:56.171874+01:00")) :metadata (message-part "{}" nil) :content (message-part "{\"name\":\"stderr\",\"text\":\"ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\\n\"}" (:name "stderr" :text "ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\n")) :buffers nil)))
  timer-event-handler([t 23677 1564 328288 nil jupyter-handle-message (#<jupyter-org-client jupyter-org-client> :iopub (:header (message-part "{\"msg_id\":\"e8c6a08d-83b1c9f7d74a3dfb1b80248e\",\"msg_type\":\"stream\",\"username\":\"boy\",\"session\":\"464ef51b-d62e8feaf3cf3b8548035101\",\"date\":\"2019-03-04T11:03:56.326894Z\",\"version\":\"5.3\"}" (:msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :username "boy" :session "464ef51b-d62e8feaf3cf3b8548035101" :date "2019-03-04T11:03:56.326894Z" :version "5.3")) :msg_id "e8c6a08d-83b1c9f7d74a3dfb1b80248e" :msg_type :stream :parent_header (message-part "{\"msg_id\":\"cfd3b78d-1ec6-409b-ba97-41c2cb82c655\",\"msg_type\":\"shutdown_request\",\"version\":\"5.3\",\"username\":\"boy\",\"session\":\"9b3dd7de-5a5f-4878-8643-a179c8a4bee2\",\"date\":\"2019-03-04T12:03:56.171874+01:00\"}" (:msg_id "cfd3b78d-1ec6-409b-ba97-41c2cb82c655" :msg_type :shutdown-request :version "5.3" :username "boy" :session "9b3dd7de-5a5f-4878-8643-a179c8a4bee2" :date "2019-03-04T12:03:56.171874+01:00")) :metadata (message-part "{}" nil) :content (message-part "{\"name\":\"stderr\",\"text\":\"ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\\n\"}" (:name "stderr" :text "ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.\n")) :buffers nil)) nil 278000])
  accept-process-output(nil 0.0001)
  jupyter-repl-restart-kernel(nil nil)
  funcall-interactively(jupyter-repl-restart-kernel nil nil)
  call-interactively(jupyter-repl-restart-kernel)
  jupyter-org--call-with-src-block-client(jupyter-repl-restart-kernel)
  jupyter-repl-restart-kernel()
  funcall-interactively(jupyter-repl-restart-kernel)
  call-interactively(jupyter-repl-restart-kernel nil nil)
  command-execute(jupyter-repl-restart-kernel)
```

Reply:
OK, I figured out the error. It has something to do with the `:shutdown-request` causing a `:stream` message to be emitted by the kernel, and the `jupyter-stream-handler` of the REPL client not handling the case when no input cell is associated with a request. Thanks for catching this, I don't think it matters that we handle any messages other than `:shutdown-reply` for a `:shutdown-request` since there would never be an association with a REPL input cell.

This normally doesn't happen for me because I have never received the error

`ERROR:root:Invalid alias: The name clear can't be aliased because it is another magic command.`

But it looks to be an issue with IPython 7.3.0 https://stackoverflow.com/questions/54971047/spyder-reports-invalid-alias-when-running-any-script, is that what you are on?

Reply:
You shouldn't get any errors now. I just ignored any other messages that are not a `:shutdown-reply` when restarting the kernel.

Reply:
That was it. The kernel is restarting as expected and all the extension functions that rely on it are working fine.